services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/model/CredentialsDto.java

```diff
@@ -0,0 +1,220 @@
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.eclipse.hono.deviceregistry.mongodb.model;
+
+import java.net.HttpURLConnection;
+import java.time.Instant;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+import org.eclipse.hono.client.ClientErrorException;
+import org.eclipse.hono.deviceregistry.mongodb.utils.MongoDbDeviceRegistryUtils;
+import org.eclipse.hono.deviceregistry.util.DeviceRegistryUtils;
+import org.eclipse.hono.service.management.credentials.CommonCredential;
+import org.eclipse.hono.service.management.credentials.CommonSecret;
+import org.eclipse.hono.service.management.credentials.GenericCredential;
+import org.eclipse.hono.service.management.credentials.PasswordCredential;
+import org.eclipse.hono.service.management.credentials.PskCredential;
+import org.eclipse.hono.service.management.credentials.X509CertificateCredential;
+import org.eclipse.hono.util.RegistryManagementConstants;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+/**
+ * A DTO (Data Transfer Object) class to store credentials information in mongodb.
+ */
+public class CredentialsDto extends BaseDto {
+
+    @JsonProperty(value = RegistryManagementConstants.FIELD_PAYLOAD_TENANT_ID, required = true)
+    private String tenantId;
+
+    @JsonProperty(value = RegistryManagementConstants.FIELD_PAYLOAD_DEVICE_ID, required = true)
+    private String deviceId;
+
+    @JsonProperty(value = MongoDbDeviceRegistryUtils.FIELD_CREDENTIALS, required = true)
+    private List<CommonCredential> credentials;
+    private boolean hasSecretIds;
+
+    /**
+     * Default constructor for serialisation/deserialization.
+     */
+    public CredentialsDto() {
+        // Explicit default constructor.
+    }
+
+    /**
+     * Creates a new data transfer object to store credentials information in mongodb.
+     *
+     * @param tenantId The tenant identifier.
+     * @param deviceId The device identifier.
+     * @param credentials The list of credentials.
+     * @param version The version of tenant to be sent as request header.
+     * @throws NullPointerException if any of the parameters except the secrets are {@code null}
+     * @throws IllegalArgumentException if validation of the given credentials fail.
+     */
+    public CredentialsDto(
+            final String tenantId,
+            final String deviceId,
+            final List<CommonCredential> credentials,
+            final String version) {
+
+        setTenantId(tenantId);
+        setDeviceId(deviceId);
+
+        //Validate the given credentials, secrets and generate secret ids if not available.
+        Optional.ofNullable(credentials)
+                .ifPresent(ok -> {
+                    validateForUniqueAuthIdAndType(credentials);
+                    validateSecretsAndGenerateIds(credentials);
+                });
```

Comment:
when will this constructor is supposed to be involved? I am asking because there seems to be an expectation for the checks to be invoked as part of certain scenarios ...

Reply:
This constructor is supposed to be involved when adding or updating credentials.